pkg/pgmodel/ingestor/copier.go

```diff
@@ -403,63 +423,96 @@ func insertSeries(ctx context.Context, conn pgxconn.PgxConn, reqs ...copyRequest
 		totalExemplars += numExemplars
 		if hasSamples {
 			numRowsPerInsert = append(numRowsPerInsert, numSamples)
-			batch.Queue("SELECT _prom_catalog.insert_metric_row($1, $2::TIMESTAMPTZ[], $3::DOUBLE PRECISION[], $4::BIGINT[])", req.info.TableName, timeSamples, valSamples, seriesIdSamples)
+			table := pgx.Identifier{req.info.TableSchema, req.info.TableName}
+			if onConflict {
+				// we append table prefix to make sure that temp table name is unique
+				table, err = createTempIngestTable(ctx, tx, req.info.TableName, req.info.TableSchema, fmt.Sprintf("s%d_", r))
+				if err != nil {
+					return err, lowestMinTime
+				}
+			}
+			inserted, err := tx.CopyFrom(ctx, table, schema.PromDataColumns[:], pgx.CopyFromRows(sampleRows))
+			if err != nil {
+				return err, lowestMinTime
+			}
+			if onConflict {
+				res, err := tx.Exec(ctx, fmt.Sprintf(sqlInsertIntoFrom, schema.PromData, pgx.Identifier{req.info.TableName}.Sanitize(),
+					strings.Join(schema.PromDataColumns[:], ","), table.Sanitize()))
+				if err != nil {
+					return err, lowestMinTime
+				}
+				inserted = res.RowsAffected()
+
+			}
+			insertedRows = append(insertedRows, int(inserted))
 		}
 		if hasExemplars {
-			// We cannot send 2-D [][]TEXT to postgres via the pgx.encoder. For this and easier querying reasons, we create a
-			// new type in postgres by the name prom_api.label_value_array and use that type as array (which forms a 2D array of TEXT)
-			// which is then used to push using the unnest method apprach.
-			labelValues := pgmodel.GetCustomType(pgmodel.LabelValueArray)
-			if err := labelValues.Set(exemplarLbls); err != nil {
-				return fmt.Errorf("setting prom_api.label_value_array[] value: %w", err), lowestMinTime
-			}
 			numRowsPerInsert = append(numRowsPerInsert, numExemplars)
-			batch.Queue("SELECT _prom_catalog.insert_exemplar_row($1::NAME, $2::TIMESTAMPTZ[], $3::BIGINT[], $4::prom_api.label_value_array[], $5::DOUBLE PRECISION[])", req.info.TableName, timeExemplars, seriesIdExemplars, labelValues, valExemplars)
+			table := pgx.Identifier{schema.PromDataExemplar, req.info.TableName}
+			if onConflict {
+				// we append table prefix to make sure that temp table name is unique
+				table, err = createTempIngestTable(ctx, tx, req.info.TableName, schema.PromDataExemplar, fmt.Sprintf("e%d_", r))
+				if err != nil {
+					return err, lowestMinTime
+				}
+			}
+			inserted, err := tx.CopyFrom(ctx, table, schema.PromExemplarColumns[:], pgx.CopyFromRows(exemplarRows))
+			if err != nil {
+				return err, lowestMinTime
+			}
+			if onConflict {
+				res, err := tx.Exec(ctx, fmt.Sprintf(sqlInsertIntoFrom, schema.PromDataExemplar, pgx.Identifier{req.info.TableName}.Sanitize(),
```

Comment:
Minor nit: if the argument list is multiline, lets put all of them on new lines. Easier to read.